File: python/1319_连通网络的操作次数_medium.py
class Solution:
    def makeConnected(self, n: int, connections: List[List[int]]) -> int:
        uf = UnionFind(n)
        count = 0
        for (x, y) in connections:
            if uf.united(x, y):
                continue
            else:
                count += 1
        return uf.setCount-1 if count>=uf.setCount-1 else -1

class UnionFind():
    def __init__(self, n):
        self.n = n
        self.setCount = n
        self.size = [1 for i in range(n)]
        self.parent = [i for i in range(n)]
    def find(self, x):
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]
    def united(self, x, y):
        x, y = self.find(x), self.find(y)
        if x == y:
            return False
        if self.size[x] < self.size[y]:
            x, y = y, x
        self.parent[y] = x
        self.size[x] += self.size[y]
        self.setCount -= 1
        return True

File: cpp/1319_连通网络的操作次数_medium.cpp
class UnionFind{
public:
    vector<int> parent;
    vector<int> size;
    int setCount;
    int n;
public:
    UnionFind(int _n): n(_n), setCount(_n), parent(_n), size(_n, 0){
        iota(parent.begin(), parent.end(), 0);
    }
    int find(int x){
        return parent[x]==x ? x : parent[x]=find(parent[x]);
    }
    bool united(int x, int y){
        x = find(x);
        y = find(y);
        if(x == y)
            return false;
        if(size[x]<size[y])
            swap(x, y);
        parent[y] = x;
        size[x] += size[y];
        setCount--;
        return true;
    }
};
class Solution {
public:
    int makeConnected(int n, vector<vector<int>>& connections) {
        UnionFind uf = UnionFind(n);
        int count = 0;
        for(auto v : connections){
            int x = v[0], y = v[1];
            if(uf.united(x, y))
                continue;
            else
                count++;
        }
        return count>=uf.setCount-1 ? uf.setCount-1 : -1;
    }
};